io/src/real_sense_2_grabber.cpp

```diff
@@ -0,0 +1,361 @@
+  {
+  }
+
+  RealSense2Grabber::~RealSense2Grabber ()
+  {
+    try
+    {
+      stop ( );
+
+      disconnect_all_slots<signal_librealsense_PointXYZ> ( );
+      disconnect_all_slots<signal_librealsense_PointXYZI> ( );
+      disconnect_all_slots<signal_librealsense_PointXYZRGB> ( );
+      disconnect_all_slots<signal_librealsense_PointXYZRGBA> ( );
+    } 
+    catch ( ... ) {}
+  }
+
+  void
+  RealSense2Grabber::start ()
+  {
+    if (isRunning ())
+      return;
+
+    // need at least one signal
+    if (signal_PointXYZ->num_slots () == 0 &&
+      signal_PointXYZI->num_slots () == 0 &&
+      signal_PointXYZRGB->num_slots () == 0 &&
+      signal_PointXYZRGBA->num_slots () == 0)
+      return;
+
+    running_ = true;
+    quit_ = false;
+
+    rs2::config cfg;
+
+    // capture from file
+    if (file_name_or_serial_number_.rfind ( ".bag" ) == file_name_or_serial_number_.length () - 4)
+    {
+      cfg.enable_device_from_file ( file_name_or_serial_number_, repeat_playback_ );
+    }
+    else
+    {
+      if (!file_name_or_serial_number_.empty ())
+        cfg.enable_device ( file_name_or_serial_number_ );
+
+      if (signal_PointXYZRGB->num_slots () > 0 || signal_PointXYZRGBA->num_slots () > 0)
+      {
+        cfg.enable_stream ( RS2_STREAM_COLOR, device_width_, device_height_, RS2_FORMAT_RGB8, target_fps_ );
+      }
+
+      cfg.enable_stream ( RS2_STREAM_DEPTH, device_width_, device_height_, RS2_FORMAT_Z16, target_fps_ );
+
+      if (signal_PointXYZI->num_slots () > 0)
+      {
+        cfg.enable_stream ( RS2_STREAM_INFRARED, device_width_, device_height_, RS2_FORMAT_Y8, target_fps_ );
+      }
+
+    }
+
+    rs2::pipeline_profile prof = pipe_.start ( cfg );
+
+    if ( prof.get_stream ( RS2_STREAM_COLOR ).format ( ) != RS2_FORMAT_RGB8 ||
+      prof.get_stream ( RS2_STREAM_DEPTH ).format ( ) != RS2_FORMAT_Z16 ||
+      prof.get_stream ( RS2_STREAM_INFRARED ).format ( ) != RS2_FORMAT_Y8 )
+      THROW_IO_EXCEPTION ( "This stream type or format not supported." );
+
+    if (signal_PointXYZRGB->num_slots () > 0 || signal_PointXYZRGBA->num_slots () > 0)
+    {
+      assert ( prof.get_stream ( RS2_STREAM_COLOR ).format () == RS2_FORMAT_RGB8 );
+    }
+
+    assert ( prof.get_stream ( RS2_STREAM_DEPTH ).format () == RS2_FORMAT_Z16 );
+
+    thread_ = std::thread ( &RealSense2Grabber::threadFunction, this );
+
+  }
+
+  void
+  RealSense2Grabber::stop ()
+  {
+    if ( !isRunning ( ) )
+      return;
+
+    running_ = false;
+    quit_ = true;
+
+    thread_.join ( );
+
+    pipe_.stop ( );
+  }
+
+  bool
+  RealSense2Grabber::isRunning () const
+  {
+    return running_;
+  }
+
+  float
+  RealSense2Grabber::getFramesPerSecond () const
+  {
+    return fps_;
+  }
+
+  void
+  RealSense2Grabber::signalsChanged ()
+  {
+    reInitialize ();
+  }
+
+  void
+  RealSense2Grabber::threadFunction ()
+  {
+    pcl::StopWatch sw;
+
+    while (!quit_)
+    {
+      sw.reset ();
+
+      rs2::depth_frame depth = nullptr;
```

Comment:
This fails to compile with GCC 7.3 on Ubuntu. 

```
error: conversion from ‘std::nullptr_t’ to non-scalar type ‘rs2::depth_frame’ requested
```

Which makes sense, how can you assign `nullptr` to a class?